doc/howtos/rdtraining/B_acl_irrules.rst

```diff
@@ -4,4 +4,424 @@
+
+* All rules which do not have any ``groups`` ("global") must match.
+* Any rule which has at least one ``groups`` ("group") can match.
+
+This means that "global" rules can only restrict (remove) access, while "group"
+rules can expand access *when added to other group rules*. However, group rules
+can not expand access beyond what the global rules allow.
+
+The result is that global rules are extremely powerful, but also difficult to
+compose as it is easy to e.g. create non-overlapping rules (which allow no
+access).
+
+Programmatic Security Checking
+==============================
+
+Access rights and access rules are only checked when *performing data access via
+the ORM* e.g. creating, reading, searching, writing, or unlinking a record via
+ORM methods. Other methods do *not* necessarily check against any sort of access
+rights.
+
+.. danger:: Just because you're already in Python code does not mean any access
+            right or rule has or will be checked.
+
+This is important and relevant when performing non-CRUD operations, or when
+wilfully bypassing the ORM for legitimate reasons.
+
+In that case, it is still possible to:
+
+* Check who the current user is, possibly match them against specific models
+  or records (``self.env.user``).
+* Check that the current user has specific groups hard-coded to allow or deny an
+  operation (``self.env.user.has_group``).
+* Call the ``check_access_rights(operation)`` method on a recorset, this
+  verifies whether the current user has access to the model itself.
+* Call ``check_access_rule(operations)`` on a non-empty recorset, this verifies
+  that the current user is allowed to perform the operation on *every* record of
+  the set.
+
+Bypassing security
+------------------
+
+There are two main ways to bypass existing security checks in Odoo, either
+wilfully or as a side-effect:
+
+* The ``sudo()`` method will create a new recorset in "sudo mode", this ignores
+  all access rules and access rights (although hard-coded group and user checks
+  may still apply).
+* Performing raw SQL queries will bypass access rules and access rights as a
+  side-effect of bypassing the ORM itself.
+
+.. danger::
+
+    These features should generally be avoided, and only used with extreme care,
+    after having checked that the current user and operation should be able to
+    bypass normal access rights validation.
+
+    Operations performed in such modes should also rely on user input as little
+    as possible, and should validate it to the maximum extent they can.
+
+.. exercise::
+
+    Having to give agents "billing" access so they can confirm a property sale
+    when ``estate_account`` is installed is giving them unnecessary access.
+
+    Update ``estate_account`` to bypass access rights and rules when creating
+    the invoice, but only if they are allowed to modify the property.
+
+.. _howto/rdtraining/B_acl_irrules/multicompany:
+
+Multi-company security
+======================
+
```

Comment:
A **Goal** where we explain that a new company is created, there is one real estate agent in each company, etc. would be useful.